core/src/main/kotlin/McuMgrResult.kt

```diff
@@ -0,0 +1,122 @@
+package com.juul.mcumgr
+
+import com.juul.mcumgr.McuMgrResult.Error
+import com.juul.mcumgr.McuMgrResult.Failure
+import com.juul.mcumgr.McuMgrResult.Success
+import com.juul.mcumgr.message.Response
+
+sealed class McuMgrResult<out T> {
+
+    data class Success<T>(
+        val value: T,
+        val code: Response.Code = Response.Code.Ok
+    ) : McuMgrResult<T>()
+
+    data class Error<T>(val code: Response.Code) : McuMgrResult<T>()
+
+    data class Failure<T>(val throwable: Throwable) : McuMgrResult<T>()
```

Comment:
It's not immediately clear what the difference is between an error vs. a failure?

Is there a need for the distinction? How would a consumer handle these differently?

Reply:
The difference is an error response vs a total failure. E.g. a response with one field `{"rc":5}` would be an error response, while a ble disconnect would be a failure.

The reason to make the distinction is because, like HTTP requests, sometimes we expect a response to return an error. E.g. when downloading a core, the device may respond with `Response.Code.NoEntry` effectively telling us that there is no core available to download. This is not a failure, nor does it conform to the expected download response format, hence being a separate entity.

For most callers, the distinction is moot. However, it would be incorrect to lump errors and failures together, since doing so may cause failures which are not truly a failure.

Reply:
Gotchya. It'd be good to name them (if possible) to be more specific.

For example, Retrofit will throw either `HttpException` which indicates that an HTTP response was received (it was just not in the 2xx range) or an `IOException` to indicate a network failure. The naming makes it clear to the library consumer what they mean and how to react to them.

Failure vs. Error doesn't provide the same meaningful naming. If a more specific naming can't be found, then KDoc should be added to them to make it clear to the library consumer what their purpose is.